typed-transitions/src/Protocol/Codec.hs

```diff
@@ -8,17 +8,122 @@
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE PolyKinds #-}
 {-# LANGUAGE DataKinds #-}
-{-# LANGUAGE GADTSyntax #-}
 {-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE GADTs #-}
 
 module Protocol.Codec where
 
-import Data.Text (Text)
-import Data.Kind (Type)
+-- | Monadic fold over an input stream with the notions of end of stream
+-- and end of fold.
+--
+-- The 'Choice' type allows the fold to express its end, and the
+-- 'Response' type forces the fold to deal with the input stream end.
+--
+-- A decoder is a `Fold concrete m (Either fail t)`.
+-- This is a sharper representation compared to 'IDecode' from cborg: if the
+-- input is exhausted (`Nothing` is given to the 'IDecode') then the decoder
+-- cannot possibly give another `Partial`; it must choose a result.
```

Comment:
Isn't that a good think: since there is no more input the decoder must decide either to fail or to return a value. In this case you eliminate partial parse after the end of input in `foldOverInput`.  So we end up with something very similar, but more complex.

Reply:
> So at the end this hides a pattern match on the Partial constructor which is an obvious error situation.

Right, this `Fold` representation doesn't allow for a partial parse at the end of input. 

>  But it's not gone: it's still in the Response type, where one needs the end which is likely to reduce to return (Left UpsErrorHappend) or something similar in this case.

That's fine, because it's not a continuation: it's either a success or a failure.